L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc

```diff
@@ -0,0 +1,1199 @@
+            float eta = getEtaMin_card(cc) * tower_width / n_crystals_towerEta - n_towers_cardEta * tower_width +
+                        (ii + 0.5) * tower_width;
+            iEta_tower_L1Card[jj][ii][cc] = getTower_absoluteEtaID(eta);
+            iPhi_tower_L1Card[jj][ii][cc] = getTower_absolutePhiID(phi);
+          }
+        }
+
+      }  // end of check if inside card
+    }    // end of loop over hits to build towers
+
+    // Loop over hcal hits to get the HCAL towers.
+    for (const auto& hit : hcalhits) {
+      if (getCrystal_phiID(hit.position().phi()) <= getPhiMax_card(cc) &&
+          getCrystal_phiID(hit.position().phi()) >= getPhiMin_card(cc) &&
+          getCrystal_etaID(hit.position().eta()) <= getEtaMax_card(cc) &&
+          getCrystal_etaID(hit.position().eta()) >= getEtaMin_card(cc) && hit.pt() > 0) {
+        for (int jj = 0; jj < n_links_card; ++jj) {
+          if ((getCrystal_phiID(hit.position().phi()) / n_crystals_towerPhi) % n_links_card == jj) {
+            for (int ii = 0; ii < n_towers_cardEta; ++ii) {
+              if ((getCrystal_etaID(hit.position().eta()) / n_crystals_towerEta) % n_towers_cardEta == ii) {
+                HCAL_tower_L1Card[jj][ii][cc] += hit.pt();
+                iEta_tower_L1Card[jj][ii][cc] = getTower_absoluteEtaID(hit.position().eta());  //hit.id().ieta();
+                iPhi_tower_L1Card[jj][ii][cc] = getTower_absolutePhiID(hit.position().phi());  //hit.id().iphi();
+              }
+            }  // end of loop over eta towers
+          }
+        }  // end of loop over phi links
+      }    // end of check if inside card
+    }      // end of loop over hits to build towers
+
+    // Give back energy of not used clusters to the towers (if there are more than 12 clusters)
+    for (unsigned int kk = n_clusters_4link; kk < cluster_list_merged[cc].size(); ++kk) {
+      if (cluster_list_merged[cc][kk].cpt > 0) {
+        ECAL_tower_L1Card[getTower_phiID(cluster_list_merged[cc][kk].cphi_)]
+                         [getTower_etaID(cluster_list_merged[cc][kk].ceta_)][cc] += cluster_list_merged[cc][kk].cpt;
+      }
+    }
+  }  //End of loop over cards
+
+  //*********************************************************
+  //******************** Do layer 2 *************************
+  //*********************************************************
+
+  // Definition of L2 outputs
+  float HCAL_tower_L2Card[n_links_GCTcard][n_towers_cardEta][3];  // 3 L2 cards send each 48 links with 17 towers
+  float ECAL_tower_L2Card[n_links_GCTcard][n_towers_cardEta][3];  // 3 L2 cards send each 48 links with 17 towers
+  int iEta_tower_L2Card[n_links_GCTcard][n_towers_cardEta][3];    // 3 L2 cards send each 48 links with 17 towers
+  int iPhi_tower_L2Card[n_links_GCTcard][n_towers_cardEta][3];    // 3 L2 cards send each 48 links with 17 towers
+  float energy_cluster_L2Card[n_links_GCTcard][2][3];             // 3 L2 cards send each 48 links with 2 clusters
```

Comment:
repeated comments:
magic numbers 2,3 should be named constants (throughout this section of code)
also, eliminate duplicate comments (just include each comment once)

Reply:
done